isisd/isis_tlvs.c

```diff
@@ -6934,3 +8160,95 @@ void isis_tlvs_set_purge_originator(struct isis_tlvs *tlvs,
+{
+	assert(!subsubtlvs->srv6_sid_structure);
+
+	subsubtlvs->srv6_sid_structure = XCALLOC(
+		MTYPE_ISIS_SUBSUBTLV, sizeof(*subsubtlvs->srv6_sid_structure));
+
+	isis_srv6_sid_structure2subsubtlv(sid, subsubtlvs->srv6_sid_structure);
+}
+
+/* Add an SRv6 End SID to the SRv6 End SID Sub-TLV */
+void isis_subtlvs_add_srv6_end_sid(struct isis_subtlvs *subtlvs,
+				   struct isis_srv6_sid *sid)
+{
+	struct isis_srv6_end_sid_subtlv *sid_subtlv;
+
+	if (!sid)
+		return;
+
+	/* The SRv6 End SID Sub-TLV advertises SRv6 SIDs with Endpoint behaviors
+	 * that do not require a particular neighbor in order to be correctly
+	 * applied (e.g. End, End.DT6, ...). Before proceeding, let's make sure
+	 * we are encoding one of the supported behaviors. */
+	if (sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END &&
+	    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_NEXT_CSID &&
+	    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT6 &&
+	    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT6_USID &&
+	    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT4 &&
+	    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT4_USID &&
+	    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT46 &&
+	    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT46_USID)
+		return;
+
+	/* Allocate memory for the Sub-TLV */
+	sid_subtlv = XCALLOC(MTYPE_ISIS_SUBTLV, sizeof(*sid_subtlv));
+
+	/* Fill in the SRv6 End SID Sub-TLV according to the SRv6 SID
+	 * configuration */
+	isis_srv6_end_sid2subtlv(sid, sid_subtlv);
+
+	/* Add the SRv6 SID Structure Sub-Sub-TLV */
+	sid_subtlv->subsubtlvs =
+		isis_alloc_subsubtlvs(ISIS_CONTEXT_SUBSUBTLV_SRV6_END_SID);
+	isis_subsubtlvs_set_srv6_sid_structure(sid_subtlv->subsubtlvs, sid);
+
+	/* Append the SRv6 End SID Sub-TLV to the Sub-TLVs list */
+	append_item(&subtlvs->srv6_end_sids, (struct isis_item *)sid_subtlv);
+}
+
+/* Add an SRv6 Locator to the SRv6 Locator TLV */
+void isis_tlvs_add_srv6_locator(struct isis_tlvs *tlvs, uint16_t mtid,
+				struct isis_srv6_locator *loc)
+{
+	bool subtlvs_present = false;
+	struct listnode *node;
+	struct isis_srv6_sid *sid;
+	struct isis_srv6_locator_tlv *loc_tlv =
+		XCALLOC(MTYPE_ISIS_TLV, sizeof(*loc_tlv));
+
+	/* Fill in the SRv6 Locator TLV according to the SRv6 Locator
+	 * configuration */
+	isis_srv6_locator2tlv(loc, loc_tlv);
+
+	/* Add the SRv6 End SID Sub-TLVs */
+	loc_tlv->subtlvs = isis_alloc_subtlvs(ISIS_CONTEXT_SUBTLV_SRV6_LOCATOR);
+	for (ALL_LIST_ELEMENTS_RO(loc->srv6_sid, node, sid)) {
+		if (sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END &&
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_NEXT_CSID &&
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT6 &&
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT6_USID &&
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT4 &&
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT4_USID &&
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT46 &&
+		    sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END_DT46_USID) {
+			isis_subtlvs_add_srv6_end_sid(loc_tlv->subtlvs, sid);
```

Comment:
Why do we call the `isis_subtlvs_add_srv6_end_sid()` when `sid->behavior != SRV6_ENDPOINT_BEHAVIOR_END...`?

Reply:
We should call `isis_subtlvs_add_srv6_end_sid()` only if `sid->behavior == SRV6_ENDPOINT_BEHAVIOR_END... ` Fixed.